Imx8mm

igoropaniuk opened the pull request:
This is rebased and adjusted version of https://github.com/foundriesio/u-boot/pull/31

Additionally switch to DM for SPL was done (for MMC/PMIC/I2C subsystems).

ldts committed:
[FIO toup] rsa-verify: fix aligment issues for SPL

Work around alignment issues when triggering verification from SPL:

Trying to boot from MMC2
sha256,rsa2048:dev"Synchronous Abort" handler, esr 0x96000021
elr: 00000000007e737c lr : 00000000007e6e78
x 0: 00000000007fd1e4 x 1: 0000000000000100
x 2: 0000000000000800 x 3: 000000000091d420
x 4: 000000000091d420 x 5: 000000000091d588
x 6: 00000000401ff9b8 x 7: 0000000000000003
x 8: 000000000091d49c x 9: 0000000000000002
x10: 00000000007fcec8 x11: 00000000307c6ebb
x12: 000000000d058cc7 x13: 00000000dfd68b5f
x14: 0000000084328582 x15: 0000000077680b62
x16: 00000000007f1cd4 x17: 0000000050808008
x18: 000000000091dea0 x19: 000000000091d520
x20: 0000000000000100 x21: 000000000091d420
x22: 00000000401ff9b8 x23: 000000000091d5b0
x24: 0000000000000100 x25: 00000000007fad58
x26: 0000000000000020 x27: 000000000091da60
x28: 0000000000000008 x29: 000000000091d3a0

Code: f94008a0 2a0103f4 aa0303f5 b40005c0 (f9400003)
Resetting CPU ...

Signed-off-by: Jorge Ramirez-Ortiz <jorge@foundries.io>
Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

ricardosalveti commented on arch/arm/mach-imx/imx8m/Kconfig:
```diff
@@ -2,8 +2,10 @@ if ARCH_IMX8M
 	select ROM_UNIFIED_SECTIONS
+	help
+	  This architecture has CAAM but we have decided to unselect
+	  it so we can enable Foundries.io secure boot (spl + itb)
```
Can you drop this extra help change? While useful, it can easily generate conflicts on rebases (just make it part of the commit message instead).

ricardosalveti commented on board/freescale/imx8mm_evk/imx8mm_evk.c:
```diff
@@ -281,9 +283,10 @@ int board_ehci_usb_phy_mode(struct udevice *dev)
 		if (state == TYPEC_STATE_SRC_RD_RA || state == TYPEC_STATE_SRC_RD)
 			return USB_INIT_HOST;
 	}
-
```
No need for this extra line removal.

ricardosalveti commented on board/freescale/imx8mm_evk/spl.c:
```diff
@@ -263,9 +135,6 @@ void board_init_f(ulong dummy)
 	enable_tzc380();
 
-	/* Adjust pmic voltage to 1.0V for 800M */
-	setup_i2c(0, CONFIG_SYS_I2C_SPEED, 0x7f, &i2c_pad_info1);
```
Won't this cause a bad side effect?

mike-scott commented on board/freescale/imx8mm_evk/spl.c:
```diff
@@ -49,174 +48,47 @@ void spl_dram_init(void)
 	ddr_init(&dram_timing);
 }
 
-#define I2C_PAD_CTRL	(PAD_CTL_DSE6 | PAD_CTL_HYS | PAD_CTL_PUE | PAD_CTL_PE)
```
Can we keep this section but protect it with something like:
```
#ifndef CONFIG_SPL_DM
```

Removing the whole block would create conflicts in future merges IMHO.

ricardosalveti commented on cmd/bootm.c:
```diff
@@ -123,7 +123,7 @@ int do_bootm(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
 	}
 
-#ifdef CONFIG_IMX_HAB
+#if defined(CONFIG_IMX_HAB) && !defined(CONFIG_IMX8M)
```
As I said on the previous PR, can you instead here create a new CONFIG flag that covers this path in a more generic way? Just because the same path will be required by other socs, and we will have to keep extending the definitions here.

A single config can easily be enable on boards that requires it, if needed.

ricardosalveti replied:
And same applies for all the other defined(CONFIG_IMX8M) lines.

ricardosalveti commented on board/freescale/imx8mm_evk/spl.c:
```diff
@@ -49,174 +48,47 @@ void spl_dram_init(void)
 	int ret;
 
-	ret = power_bd71837_init(I2C_PMIC);
+	/*ret = power_bd71837_init(I2C_PMIC);
```
Is this replaced by CONFIG_SPL_DM_PMIC_BD7183? If so, we can probably just remove the lines instead.

ricardosalveti reviewed (changes requested):
Added some minor comments, mostly from the previous PR.

mike-scott commented on cmd/bootm.c:
```diff
@@ -123,7 +123,7 @@ int do_bootm(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
 	}
 
-#ifdef CONFIG_IMX_HAB
+#if defined(CONFIG_IMX_HAB) && !defined(CONFIG_IMX8M)
```
This normally doesn't need to be ifdef'd out.
I wonder if this isn't a side-effect of having `CONFIG_IMX_OPTEE` enabled.  We don't actually support that boot flow (loading OPTEE from u-boot vs. SPL).

igoropaniuk committed:
[FIO internal] imx8m: don't enable CAAM if OP-TEE is used

Don't enable CAAM in U-Boot, if OP-TEE is used.

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

igoropaniuk committed:
[FIO internal] usb hub: don't use env if it's not enabled

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

igoropaniuk commented:
@mike-scott @ricardosalveti pushed updated version, currently fighting with:
```
Unfortunately SPL doesn't fit in SRAM:
aarch64-linux-gnu-ld.bfd: u-boot-spl section `.data' will not fit in region `.sram'
aarch64-linux-gnu-ld.bfd: region `.sram' overflowed by 4152 bytes

```
when SPL_USB is enabled

ricardosalveti commented:
Thanks, looking a lot better now.

We might be able to drop SPL_USB and load everything to ram using the ROM SDP (SPL + u-boot.itb), as done on imx6ullevk.

e.g.
```
SDP: dcd -f SPL-mfgtool
SDP: write -f SPL-mfgtool -ivt 0
SDP: write -f u-boot-mfgtool.itb -addr 0x80800000
SDP: jump -f SPL-mfgtool -ivt
```

igoropaniuk commented on cmd/Makefile:
```diff
@@ -191,7 +191,11 @@ endif # !CONFIG_SPL_BUILD
 obj-$(CONFIG_$(SPL_)CMD_TLV_EEPROM) += tlv_eeprom.o
 
 # core command
+ifndef CONFIG_SPL_BUILD
```
this change looks pretty suspucious (however it fixed linking issues).
Maybe I've broken something that caused nvedit compilation for SPL even with disabled ENV support (`CONFIG_SPL_ENV_SUPPORT`)

mike-scott replied:
We shouldn't need this as a compile fix.  Please re-test w/ the config fragments we've shared.

igoropaniuk commented:
Droped two commits that were chaging default defconfig for imx8mm_evk, move that changes to lmp.cfg instead

igoropaniuk committed:
[FIO internal] imx8mm_evk: fix build issues when SPL_DM=y

1. Force disable DM_I2C for the SPL build (workaround)
2. Disable usage of TCPC from the board file for the SPL build, as
from board/freescale/common/Makefile:

ifndef CONFIG_SPL_BUILD
obj-$(CONFIG_USB_TCPC) += tcpc.o
endif

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

mike-scott commented on arch/arm/mach-imx/imx8m/Kconfig:
```diff
@@ -2,7 +2,7 @@ if ARCH_IMX8M
 config IMX8M
 	bool
-	select HAS_CAAM
+	select HAS_CAAM if !IMX_OPTEE
```
Can we change this to:
```
select HAS_CAAM if !OPTEE
```

mike-scott commented on lib/rsa/rsa-verify.c:
```diff
@@ -321,11 +321,12 @@ static int rsa_verify_key(struct image_sign_info *info,
 	if (ret) {
-		printf("RSA: Can't find Modular Exp implementation\n");
-		return -EINVAL;
+		printf("RSA: No Modular Exp HW implementation, use SW\n");
```
This code should probably be straightened out upstream.  The check for `USE_HOSTCC` is very odd.  However, we shouldn't hit a problem in this code as it is, because we are using `SPL_DM`.

mike-scott commented on lib/rsa/rsa-verify.c:
```diff
@@ -380,6 +384,11 @@ static int rsa_verify_with_keynode(struct image_sign_info *info,
 	if (!prop.public_exponent || length < sizeof(uint64_t))
 		prop.public_exponent = NULL;
 
+#if defined(CONFIG_SPL_RSA)
```
Let's get some more testing on this code to verify that we need this fix.

igoropaniuk replied:
that patch for aligment work around is still needed, otherwise:

```
   Trying to boot from MMC2
    sha256,rsa2048:dev"Synchronous Abort" handler, esr 0x96000021
    elr: 00000000007e737c lr : 00000000007e6e78
    x 0: 00000000007fd1e4 x 1: 0000000000000100
    x 2: 0000000000000800 x 3: 000000000091d420
    x 4: 000000000091d420 x 5: 000000000091d588
    x 6: 00000000401ff9b8 x 7: 0000000000000003
    x 8: 000000000091d49c x 9: 0000000000000002
    x10: 00000000007fcec8 x11: 00000000307c6ebb
    x12: 000000000d058cc7 x13: 00000000dfd68b5f
    x14: 0000000084328582 x15: 0000000077680b62
    x16: 00000000007f1cd4 x17: 0000000050808008
    x18: 000000000091dea0 x19: 000000000091d520
    x20: 0000000000000100 x21: 000000000091d420
    x22: 00000000401ff9b8 x23: 000000000091d5b0
    x24: 0000000000000100 x25: 00000000007fad58
    x26: 0000000000000020 x27: 000000000091da60
    x28: 0000000000000008 x29: 000000000091d3a0
    Code: f94008a0 2a0103f4 aa0303f5 b40005c0 (f9400003)
    Resetting CPU ...
```

mike-scott commented on arch/arm/mach-imx/Kconfig:
```diff
@@ -87,6 +87,14 @@ config SPL_IMX_HAB
 	  This option enables the support for secure boot (HAB) in SPL.
 	  See doc/README.mxc_hab for more details.
 
+config IMX_HAB_IMAGE_AUTH
```
I would rather see this commit be submitted upstream.  In general, this is not our use-case as we avoid the entire code path by not enabling `IMX_OPTEE`.

mike-scott replied:
And we don't use `booti` / `bootz` code paths.

igoropaniuk committed:
[FIO internal] imx8mmevk: drop hardcoded USB configuration

Drop hardcoded USB configuration from imx8mmevk header.

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

mike-scott reviewed (approved):
I've been testing this PR + https://github.com/foundriesio/u-boot/pull/41 and for me this looks good.

ricardosalveti reviewed (approved):
LGTM
